projects/cdk/directives/auto-focus/autofocus.directive.ts

```diff
@@ -10,25 +10,31 @@ import {
     Self,
 } from '@angular/core';
 import {ANIMATION_FRAME} from '@ng-web-apis/common';
-import {POLLING_TIME} from '@taiga-ui/cdk/constants';
-import {
-    TuiFocusableElementAccessor,
-    TuiNativeFocusableElement,
-} from '@taiga-ui/cdk/interfaces';
+import {TuiFocusableElementAccessor} from '@taiga-ui/cdk/interfaces';
 import {TUI_FOCUSABLE_ITEM_ACCESSOR, TUI_IS_IOS} from '@taiga-ui/cdk/tokens';
-import {setNativeFocused} from '@taiga-ui/cdk/utils/focus';
-import {Observable, race, timer} from 'rxjs';
-import {map, skipWhile, take, throttleTime} from 'rxjs/operators';
+import {Observable} from 'rxjs';
 
-const TIMEOUT = 1000;
-const NG_ANIMATION_SELECTOR = '.ng-animating';
+import {TuiAutofocusHandler} from './autofocus.options';
+import {TuiDefaultAutofocusHandler} from './handlers/default.handler';
+import {TuiIosAutofocusHandler} from './handlers/ios.handler';
 
 // TODO: 3.0 change input name to tuiAutoFocus and handle empty string
 // TODO: refactor on this whole thing in 3.0
-@Directive({
-    selector: '[tuiAutoFocus]',
-})
+@Directive({selector: '[tuiAutoFocus]'})
 export class TuiAutoFocusDirective implements AfterViewInit {
+    private readonly handler: TuiAutofocusHandler = this.isIos
+        ? new TuiIosAutofocusHandler(
+              this.tuiFocusableComponent,
+              this.elementRef,
+              this.renderer,
+              this.ngZone,
+          )
+        : new TuiDefaultAutofocusHandler(
+              this.tuiFocusableComponent,
```

Comment:
I wish there was a way to use `useClass` so we don't have to know all the dependencies. Maybe just for this situation we can reach for global variable? Something like:
```ts
{
  provider: AbstractTuiAutofocusHandler,
  useClass: typeof navigator !== 'undefined' && isApple(navigator)
    ? TuiIosAutofocusHandler
    : TuiDefaultAutofocusHandler
}
```

Reply:
@waterplea sorry, I have this
<img width="663" alt="image" src="https://user-images.githubusercontent.com/12021443/169248426-77adf0c8-acf3-49a3-bda6-1bceb1ca980a.png">

Reply:
@waterplea sorry, I have this
<img width="663" alt="image" src="https://user-images.githubusercontent.com/12021443/169248426-77adf0c8-acf3-49a3-bda6-1bceb1ca980a.png">

Reply:
fixed